### PR TITLE
Edited Nodal Kernels to be boundary restrictable

### LINKED ISSUE #6057: Edit FEProblem to allow userobjects to make changes the residual vector
In FEProblem:computeResidualType the user objects are executed first and then Nonlinearsystem::computeResidual is called. The residual vector is initialized to zero inside Nonlinearsystem::computeResidual so any changes made to the residual vector by the user objects is lost. Initializing the residual vector to zero before executing post aux user objects solves this issue.

### PR DESCRIPTION
**Pull request by sveerara**
Following the comments by @andrsd and @friedmud , I have edited Nodal Kernels so that they can be restricted to a boundary. Thanks!

Closes #6057

### CODE REVIEW

File: framework/include/base/ComputeNodalKernelBCJacobiansThread.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+// libMesh includes
+#include "libmesh/node_range.h"
+#include "libmesh/numeric_vector.h"
+
+
+class FEProblem;
+class AuxiliarySystem;
+
+
+class ComputeNodalKernelBCJacobiansThread
```
Comment (andrsd) [inline framework/include/base/ComputeNodalKernelBCJacobiansThread.h]: Should not this class inherit from `ThreadedNodeLoop<ConstNodeRange, ConstNodeRange::const_iterator>` ?

File: framework/include/base/ComputeNodalKernelBcsThread.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+#include "ParallelUniqueId.h"
+#include "NodalKernelWarehouse.h"
+#include "MooseMesh.h"
+
+class FEProblem;
+class AuxiliarySystem;
+
+
+class ComputeNodalKernelBcsThread
```
Comment (andrsd) [inline framework/include/base/ComputeNodalKernelBcsThread.h]: Should not this class inherit from `ThreadedNodeLoop<ConstNodeRange, ConstNodeRange::const_iterator>` ?

File: framework/include/base/ComputeNodalKernelBcsThread.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+{
+public:
+  ComputeNodalKernelBcsThread(FEProblem & fe_problem, AuxiliarySystem & sys, std::vector<NodalKernelWarehouse> & nodal_kernels);
+  // Splitting Constructor
+  ComputeNodalKernelBcsThread(ComputeNodalKernelBcsThread & x, Threads::split split);
+
+  //void operator() (const ConstBndNodeRange & range);
```
Comment (andrsd) [inline framework/include/base/ComputeNodalKernelBcsThread.h]: Remove this line

File: framework/include/base/ComputeNodalKernelBCJacobiansThread.h (hunk 1 above)
Comment (andrsd) [inline framework/include/base/ComputeNodalKernelBCJacobiansThread.h]: The type should actually be `ConstBndNodeRange`, not `ConstNodeRange`.

File: framework/include/base/ComputeNodalKernelBcsThread.h (hunk 2 above)
Comment (andrsd) [inline framework/include/base/ComputeNodalKernelBcsThread.h]: Same here, the type should actually be `ConstBndNodeRange`.

File: framework/src/base/NonlinearSystem.C [hunk 4]
```diff
@@ -1265,6 +1271,25 @@ NonlinearSystem::computeResidualInternal(Moose::KernelType type)
...
+    if (!_nodal_kernels[0].allBoundaryNodalKernels().empty())
+    {
+      ComputeNodalKernelBcsThread cnk(_fe_problem, _fe_problem.getAuxiliarySystem(), _nodal_kernels);
+
+      ConstBndNodeRange & bnd_node_range = *_mesh.getBoundaryNodeRange();
+
+      Threads::parallel_reduce(bnd_node_range, cnk);
+
+      unsigned int n_threads = libMesh::n_threads();
+      for (unsigned int i=0; i<n_threads; i++) // Add any cached residuals that might be hanging around
```
Comment (andrsd) [inline framework/src/base/NonlinearSystem.C]: This should have operators '=' and  '<' padded with spaces.

File: framework/src/base/NonlinearSystem.C [hunk 5]
```diff
@@ -1843,6 +1869,18 @@ NonlinearSystem::computeJacobianInternal(SparseMatrix<Number> &  jacobian)
...
+
+        // Boundary restricted Nodal Kernels
+        if (!_nodal_kernels[0].allBoundaryNodalKernels().empty())
+        {
+          ComputeNodalKernelBCJacobiansThread cnkjt(_fe_problem, _fe_problem.getAuxiliarySystem(), _nodal_kernels, jacobian);
+          ConstBndNodeRange & bnd_range = *_mesh.getBoundaryNodeRange();
+
+          Threads::parallel_reduce(bnd_range, cnkjt);
+          unsigned int n_threads = libMesh::n_threads();
+          for (unsigned int i=0; i<n_threads; i++) // Add any cached jacobians that might be hanging around
```
Comment (andrsd) [inline framework/src/base/NonlinearSystem.C]: This should have operators '=' and  '<' padded with spaces.

File: framework/src/base/NonlinearSystem.C [hunk 6]
```diff
@@ -1866,6 +1905,19 @@ NonlinearSystem::computeJacobianInternal(SparseMatrix<Number> &  jacobian)
...
+        // Boundary restricted Nodal Kernels
+        if (!_nodal_kernels[0].allBoundaryNodalKernels().empty())
+        {
+          ComputeNodalKernelBCJacobiansThread cnkjt(_fe_problem, _fe_problem.getAuxiliarySystem(), _nodal_kernels, jacobian);
+          ConstBndNodeRange & bnd_range = *_mesh.getBoundaryNodeRange();
+
+          Threads::parallel_reduce(bnd_range, cnkjt);
+
+          unsigned int n_threads = libMesh::n_threads();
+          for (unsigned int i=0; i<n_threads; i++) // Add any cached jacobians that might be hanging around
```
Comment (andrsd) [inline framework/src/base/NonlinearSystem.C]: This should have operators '=' and  '<' padded with spaces.

File: framework/src/nodalkernels/NodalKernelWarehouse.C [hunk 7]
```diff
@@ -60,15 +60,35 @@ NodalKernelWarehouse::addNodalKernel(MooseSharedPointer<NodalKernel> & nodal_ker
...
+    _all_boundary_nodal_kernels.push_back(nodal_kernel);
+
+    // Get the BoundaryIDs for this object
+    const std::set<BoundaryID> & boundary_ids = nodal_kernel->boundaryIDs();
+
+    // Populate the elemental and nodal block restricted maps
+    for (std::set<BoundaryID>::const_iterator it = boundary_ids.begin(); it != boundary_ids.end(); ++it)
+      _active_boundary_nodal_kernels[*it].push_back(nodal_kernel);
+  }
+
```
Comment (andrsd) [inline framework/src/nodalkernels/NodalKernelWarehouse.C]: Do not need an empty line here...

File: framework/src/base/ComputeNodalKernelBcsThread.C [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+#include "FEProblem.h"
+#include "NodalKernel.h"
+#include "NodalKernelWarehouse.h"
+
+// libmesh includes
+#include "libmesh/threads.h"
+
+ComputeNodalKernelBcsThread::ComputeNodalKernelBcsThread(FEProblem & fe_problem,
+                                                   AuxiliarySystem & sys,
+                                                   std::vector<NodalKernelWarehouse> & nodal_kernels) :
```
Comment (andrsd) [inline framework/src/base/ComputeNodalKernelBcsThread.C]: Weird indentation. Either put all on one line or indent on types.

File: framework/src/base/ComputeNodalKernelBCJacobiansThread.C [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+#include "NodalKernel.h"
+#include "NodalKernelWarehouse.h"
+
+// libmesh includes
+#include "libmesh/threads.h"
+
+ComputeNodalKernelBCJacobiansThread::ComputeNodalKernelBCJacobiansThread(FEProblem & fe_problem,
+                                                                     AuxiliarySystem & sys,
+                                                                     std::vector<NodalKernelWarehouse> & nodal_kernels,
+                                                                     SparseMatrix<Number> & jacobian) :
```
Comment (andrsd) [inline framework/src/base/ComputeNodalKernelBCJacobiansThread.C]: Weird indentation. Either put all on one line or indent on types.

**Comment by sveerara** [comment]
I have addressed the comments. Thanks!

File: framework/src/base/NonlinearSystem.C [hunk 10]
```diff
@@ -1259,13 +1265,32 @@ NonlinearSystem::computeResidualInternal(Moose::KernelType type)
...
+
+      unsigned int n_threads = libMesh::n_threads();
+      for (unsigned int i = 0; i < n_threads; i++) // Add any cached residuals that might be hanging around
         _fe_problem.addCachedResidual(i);
     }
   }
   PARALLEL_CATCH;
 
   if (_need_residual_copy)
+  if (_need_residual_copy)
```
Comment (andrsd) [inline framework/src/base/NonlinearSystem.C]: One if-statement is enough ;-)

File: framework/src/base/ComputeNodalKernelBCJacobiansThread.C [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+          for (std::vector<MooseSharedPointer<NodalKernel> >::iterator nodal_kernel_it = active_involved_kernels.begin();
+               nodal_kernel_it != active_involved_kernels.end();
+               ++nodal_kernel_it)
+            (*nodal_kernel_it)->computeOffDiagJacobian(jvar);
+
+          _num_cached++;
+        }
+      }
+
+      if (_num_cached == 20) //cahce 20 nodes worth before adding into the jacobian
```
Comment (andrsd) [inline framework/src/base/ComputeNodalKernelBCJacobiansThread.C]: `cahce` => `cache`

**Comment by sveerara** [comment]
Corrected typo and removed extra if statement. Thanks!

**Comment by andrsd** [comment]
:+1:

**Pull request merged**